### ISSUE TITLE
Take user-friendly measures in the event of jslib load failure

### ISSUE THREAD

**Issue by frank-emrich** [links-lang/links#590]
I recently had a case where Sam and me thought that the examples were broken, when actually I had just entered a path for jslib in my config file that was not accessible. The webserver should simply not start in this case.

**Comment by jamescheney** [comment]
Good idea, it is a huge pain when this silently fails.  (I've learned through bitter experience to check the js console...)

**Comment by dhil** [comment]
The webserver does not know about `jslib`. The webserver only attempts to serve requests, and one can request anything. We **do not want** the webserver to stop if it could not locate, say, `jslib.js`. Because anyone could attack/disrupt the server by simply by requesting a non-existent file.

**Comment by dhil** [comment]
A better approach may be to modify the page compiler to insert a check to see whether `jslib.js` has been loaded, and if it has not then produce a error page.

**Comment by frank-emrich** [comment]
> The webserver does not know about jslib. The webserver only attempts to serve requests, and one can request anything. We do not want the webserver to stop if it could not locate, say, jslib.js

It does, however, know about `jslibdir`. Thus, it could simply reject starting up when that particular directory doesn't exist, which seems justified to me. This does not catch the case where the `jslib.js` file is missing, but I feel like that's the less likely error.

I guess your solution shouldn't be much more difficult to implement. Its only downside is that you don't see the error until you request a page.

**Comment by jstolarek** [comment]
I'm very much in support of having some sort of warning of the `jslibdir` does not point to a valid directory. It took my around 10 minutes to figure out that `jslibdir` has to be an absolute path rather than a relative one. (Or perhaps it can be a relative one but I couldn't figure out what it's supposed to be.) Perhaps we could have a warning at startup displayed at the console?

**Comment by jamescheney** [comment]
One problem with this is that it's perfectly fine to run Links without interactively without `jslibdir`.  So we need to determine somehow whether Links is (going to be0 running web mode.  We used to have a specific flag for this but currently it just depends on whether `servePages` gets called.  So perhaps we could modify `servePages` to do this sanity check and fail if `jslibdir` doesn't resolve to a sensible path.  Or perhaps we could check that IF `jslibdir` is provided as a config option then it must resolve to a sensible path, but if not we assume the user knows what they are doing and isn't going to try to run in web mode.

Any preferences among these two (or am I missing something and it's simpler than I think)?

**Comment by ezrakilty** [comment]
I can see there are several ways to signal the error for this.

When I hit this, my expectation was that it would serve an error page for the request to /lib/jslib.js, and log said error to the console. I think I would have realized my mistake quickly in that case.

A good comparison might be a web framework like Django. When I run a Django development server on the command line, it starts emitting log lines for all requests (like an Apache access.log) to stdout. These give me confidence that I'm actually hitting that server in the successful case, and in the unsuccessful case, I can see the HTTP errors (400 for Bad request, say) in the output. (The exact logging behavior is controlled by a settings file, since in production you want to send it to a log file rather than stdout.)

I think the servePages feature of Links is really a "web framework" along the lines of Django or something, and can probably take a page (haha) from that book.

**Comment by jamescheney** [comment]
That makes sense to me.  I think this specific issue is a common enough case that it still makes sense to have a specific fix for it, but general web server / framework style logging seems like a sensible feature request.

**Comment by ezrakilty** [comment]
To be clear, I was riffing, but I think the key fix is to return a 400 error for any unrecognized URL.

Secondarily, I think `servePages` could crash the server if it is called when `jslibdir` is not set. This would crash on startup (not in response to an attacker's request), but only if `servePages` is used, not for REPL invocations or non-web programs.

**Comment by dhil** [comment]
I think the lowest effort, most forward-compatible fix to this issue would be to instrument the Links JavaScript runtime to check whether the required dependencies are loaded (nowadays that's only `jslib.js`). If they aren't, then the runtime should halt and error before attempting to run the initialisation procedure.

It is straightforward to implement this logic in our runtime, the only slightly hacky thing we would need is to include "magic" constants in each dependency. The runtime needs to check whether each of these constants are defined and have the proper "magic value".

For example the `jslib` might define a constant `const _JS_LIB = "foo bar baz";` (the underscore makes it inaccessible in Links source code, although, you could expose it indirectly via the alien interface), then the runtime simply need to run code akin to the following during startup
```javascript
if (_JS_LIB === void 0 || _JS_LIB !== "magic value") error;
```
(`void 0` is the canonical value for `undefined`.) The error procedure could include typical diagnostics.

**Comment by jamescheney** [comment]
Is this fixed by #700 or do we still want to do something to improve error messages here?

**Comment by SimonJF** [comment]
I think #700 fixes the most common cause of the error, but this issue pertains more to reporting the error message (should it arise through other misconfigurations). So I think we should keep it open (at least until I actually address it as per @dhil's proposal, probably later today).
